dali/kernels/reduce/reduce_gpu_impl.cuh

```diff
@@ -0,0 +1,1092 @@
+  template <typename T>
+  void AddIO(int nsamples, size_t alignment = alignof(T)) {
+    io_buffers = align_up(io_buffers, alignment) + nsamples * sizeof(T);
+  }
+};
+
+
+/**
+ * @brief Manages working memory for multi-stage operations
+ *
+ * The layout is:
+ * Host:  parameters
+ * GPU:   parameters  I/O ......... O/I
+ *
+ * Every other stage the roles of inputs and outputs are switched, so the outputs of the
+ * previous stage become the inputs of current one.
+ */
+struct WorkArea {
+  char *host_memory = nullptr, *gpu_memory = nullptr;
+
+  int64_t next_param_offset = 0;
+  int stage_index = 0;
+
+  void BeginStage(int stage_idx) {
+    next_param_offset = 0;
+    stage_index = stage_idx;
+  }
+
+  template <typename T>
+  T *ParamBuffer(int64_t n, size_t alignment = alignof(T)) {
+    next_param_offset = align_up(next_param_offset, alignment);
+    T *out = reinterpret_cast<T *>(host_memory + next_param_offset);
+    next_param_offset += n * sizeof(T);
+    assert(next_param_offset <= buffer_sizes.param_buffers);
+    return out;
+  }
+
+  template <typename T>
+  const T *InputBuffer(int64_t elements) {
+    return IOBuffer<const T>(elements, (stage_index & 1) == 1);
+  }
+
+  template <typename T>
+  T *OutputBuffer(int64_t elements) {
+    return IOBuffer<T>(elements, (stage_index & 1) == 0);
+  }
+
+  template <typename T>
+  T *IOBuffer(int64_t elements, bool alloc_at_end) {
+    int64_t offset;
+    if (alloc_at_end) {
+      offset = buffer_sizes.io_buffers + buffer_sizes.param_buffers - elements * sizeof(T);
+      offset &= -alignof(T);  // we're allocating from the end, so align _down_
+    } else {
+      offset = buffer_sizes.param_buffers;
+      offset = align_up(offset, alignof(T));
+    }
+    return reinterpret_cast<T *>(gpu_memory + offset);
+  }
+
+  void CopyParamsToDevice(cudaStream_t stream) {
+    if (next_param_offset) {
+      CUDA_CALL(cudaMemcpyAsync(gpu_memory, host_memory, next_param_offset,
+                                cudaMemcpyHostToDevice, stream));
+    }
+  }
+
+  template <typename T>
+  T *GetDeviceParam(T *host_param) const {
+    if (host_param == nullptr)
+      return nullptr;
+
+    auto raw_ptr = reinterpret_cast<const char*>(host_param);
+    assert(raw_ptr >= host_memory &&
+           raw_ptr < host_memory + next_param_offset);
+    return reinterpret_cast<T*>(gpu_memory + (raw_ptr - host_memory));
+  }
+
+  TempBufferSizes buffer_sizes;
+};
+
+
+template <typename Out, typename In, typename Actual>
+class ReduceImplGPU {
+ public:
+  KernelRequirements Setup(KernelContext &ctx,
+                           const TensorListShape<> &in_shape,
+                           span<const int> axes,
+                           bool keep_dims,
+                           bool reduce_batch) {
+    if (in_shape.sample_dim() > 64)
+      throw std::range_error("Reduce supports up to 64 dimensions");
+    reduce_batch_ = reduce_batch;
+    CheckAxes(axes, in_shape.sample_dim());
+    if (reduce_batch)
```

Comment:
```suggestion
    if (reduce_batch_)
```

Reply:
I can do it, but using the parameter is safer, in case the check is moved above the member assignment.